tools/rpkg/src/scan.cpp

```diff
@@ -6,20 +6,262 @@
+		return (data_ptr_t)INTEGER_POINTER(coldata);
+	case RType::NUMERIC:
+		return (data_ptr_t)NUMERIC_POINTER(coldata);
+	case RType::INTEGER64:
+		return (data_ptr_t)NUMERIC_POINTER(coldata);
+	case RTypeId::FACTOR:
+		// TODO What about factors that use numeric?
+		return (data_ptr_t)INTEGER_POINTER(coldata);
+	case RType::STRING:
+		return (data_ptr_t)DATAPTR_RO(coldata);
+	case RType::TIMESTAMP:
+		return (data_ptr_t)NUMERIC_POINTER(coldata);
+	case RType::TIME_SECONDS:
+	case RType::TIME_MINUTES:
+	case RType::TIME_HOURS:
+	case RType::TIME_DAYS:
+	case RType::TIME_WEEKS:
+		return (data_ptr_t)NUMERIC_POINTER(coldata);
+	case RType::TIME_SECONDS_INTEGER:
+	case RType::TIME_MINUTES_INTEGER:
+	case RType::TIME_HOURS_INTEGER:
+	case RType::TIME_DAYS_INTEGER:
+	case RType::TIME_WEEKS_INTEGER:
+		return (data_ptr_t)INTEGER_POINTER(coldata);
+	case RType::DATE:
+		if (!IS_NUMERIC(coldata)) {
+			cpp11::stop("DATE should really be integer");
+		}
+		return (data_ptr_t)NUMERIC_POINTER(coldata);
+	case RType::DATE_INTEGER:
+		if (!IS_INTEGER(coldata)) {
+			cpp11::stop("DATE_INTEGER should really be integer");
+		}
+		return (data_ptr_t)INTEGER_POINTER(coldata);
+	case RType::LIST_OF_NULLS:
+	case RType::BLOB:
+		return (data_ptr_t)DATAPTR_RO(coldata);
+	case RTypeId::LIST:
+		return (data_ptr_t)DATAPTR_RO(coldata);
+	case RTypeId::STRUCT:
+		// Will bind child columns dynamically. Could also optimize by descending early and recording.
+		return (data_ptr_t)coldata;
+	default:
+		cpp11::stop("rapi_execute: Unsupported column type for bind");
+	}
+}
+
+struct DedupPointerEnumType {
+	static bool IsNull(SEXP val) {
+		return val == NA_STRING;
+	}
+	static uintptr_t Convert(SEXP val) {
+		return (uintptr_t)DATAPTR(val);
+	}
+};
+
 template <class SRC, class DST, class RTYPE>
-static void AppendColumnSegment(SRC *source_data, Vector &result, idx_t count) {
-	auto result_data = FlatVector::GetData<DST>(result);
+static void AppendColumnSegment(SRC *source_data, idx_t sexp_offset, Vector &result, idx_t count) {
+	source_data += sexp_offset;
 	auto &result_mask = FlatVector::Validity(result);
 	for (idx_t i = 0; i < count; i++) {
 		auto val = source_data[i];
 		if (RTYPE::IsNull(val)) {
 			result_mask.SetInvalid(i);
 		} else {
+			auto result_data = FlatVector::GetData<DST>(result);
 			result_data[i] = RTYPE::Convert(val);
 		}
 	}
 }
 
+void AppendListColumnSegment(const RType &rtype, SEXP *source_data, idx_t sexp_offset, Vector &result, idx_t count) {
+	source_data += sexp_offset;
+	auto &result_mask = FlatVector::Validity(result);
+	auto child_rtype = rtype.GetListChildType();
+	for (idx_t i = 0; i < count; i++) {
+		auto val = source_data[i];
+		if (RSexpType::IsNull(val)) {
+			result_mask.SetInvalid(i);
+		} else {
+			auto result_data = FlatVector::GetData<list_entry_t>(result);
+			auto len = RApiTypes::GetVecSize(child_rtype, val);
+			result_data[i].offset = ListVector::GetListSize(result);
+			for (R_len_t i = 0; i < len; i++) {
```

Comment:
Just for my understanding, we are just pushing back the nested lists, but not every element in the list?

Reply:
I'll rename the `i`. If this doesn't help, can you please rephrase?

Reply:
never mind, I think I figured it out